Docs/hashicorp key management.md (#114)
mariadb-corporation/mariadb-docs: tools/mariadb-enterprise-operator/plugins/hashicorp-key-management.md

- Vault is unsealed and you've logged in to it with `vault login $AUTH_TOKEN`, where $AUTH_TOKEN is an authentication token given to you by an administrator
- (Optional) `openssl` for generating secrets

### Steps 

1. **Creating A New Key-Value Store In Vault.**
    Create a new key-value store and take note of the `path`. In our example we will use `mariadb`.
    ```sh
    vault secrets enable -path /mariadb -version=2 kv
    ```

2. **Adding necessary secrets.**
    We will put 2 secrets with ids `1` and `2`. `2` will be used for temporary files, while `1` will be used for everything else. It is not
    neccessary to create 2 of them and in that case, temporary files will use `1`.

    Note: Here you should use the `path` we chose in the previous step.
    ```sh
    vault kv put /mariadb/1 data="$(openssl rand -hex 32)"
    vault kv put /mariadb/2 data="$(openssl rand -hex 32)"
    ```

3. **(Optional) Create An Authentication Token.**
    This step can be skipped if you want to use your own token. Consult with a Vault administrator regarding this. If you want to create a new one:
    ```sh

>     vault token create

Reviewer: Although the [plugin documentation](https://mariadb.com/docs/server/security/securing-mariadb/encryption/data-at-rest-encryption/key-management-and-encryption-plugins/hashicorp-key-management-plugin) says:
>  Authentication is done using the Hashicorp Vault's token authentication method

This will be a static token, something not desirable by any security team. We could stick to this docs, but refer to the other available Vault authentication methods, which allow you to generate a short-lived Vault token. For example, the approle one can generate a Vault token based on a acces id and secret key:
https://developer.hashicorp.com/vault/docs/auth/approle

So the user could authenticate with any of the [supported methods by Vault](https://developer.hashicorp.com/vault/docs/auth), and obtain a Vault token, rather than creating a static token.

Reviewer 2: approle will be a problem too, unless it has 0 ttl, which wouldn't really be any different from tokens I believe.

    ```
    You will see output similar to:
    ```
    Key                  Value
    ---                  -----
    token                EXAMPLE_TOKEN
    token_accessor       9sHvwKiddpCT2PaplcHHRAnv
    token_duration       ∞
    token_renewable      false

>     token_policies       ["root"]

Reviewer: Ideally, we should have a custom policy that allows all operations under the `/mariadb/*` keys, rather than in the entire Vault.

Reviewer: I know this is done like this [in the blog](https://mariadb.com/resources/blog/how-to-implement-encryption-at-rest-using-hashicorp-vault-and-mariadb/), but it is far from ideal.

Besides, the plugin is not configuring any authentication method nor anything global in vault, it just reads keys under a kv path, so their permissions should be scoped to `/mariadb/*`.

Author: Yeah, the dev server didn't care for policies :smile: But now since we want a TLS example I had to spin up a standalone instance and a custom policy was a must.

    identity_policies    []
    policies             ["root"]
    ```
    Your new token is: `EXAMPLE_TOKEN`.

4. **Create A `ConfigMap` For my.cnf**
    Now, because we will store a Vault authentication token in `my.cnf`, to facilitate encryption and gitops, we will put the `my.cnf` configuration in a separate `ConfigMap` instead of having it directly as
    part of the spec. This way, you can encrypt the `ConfigMap` only and store it in git, or even create the `ConfigMap` with an external tool.

    **mariadb-vault-config.yaml**
    ```yaml
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: mariadb-vault
      labels:
        enterprise.mariadb.com/watch: ""
    data:
      my.cnf: |
        [mariadb]
        bind-address=*
        default_storage_engine=InnoDB
        binlog_format=row
        innodb_autoinc_lock_mode=2
        innodb_buffer_pool_size=800M
        max_allowed_packet=256M

        plugin_load_add = hashicorp_key_management

>         hashicorp-key-management-vault-url=http://vault-0.vault-internal.vault.svc.cluster.local:8200/v1/mariadb # Change to your Vault URL.

Reviewer: We should include a `hashicorp-key-management-vault-ca` to establish the TLS trust:
https://mariadb.com/docs/server/security/securing-mariadb/encryption/data-at-rest-encryption/key-management-and-encryption-plugins/hashicorp-key-management-plugin#hashicorp-key-management-vault-ca

>         hashicorp-key-management-token=EXAMPLE_TOKEN # This needs to be changed to your token

Reviewer: ⚠️ We cannot  seed a Vault token in a ConfigMap, specially if it has elevated permissions

Alternatively, we can supply the HashiCorp configuration in a `Secret`,  and use the `volumes` and `volumeMounts` to mount it under ` /etc/mysql/mariadb.conf.d/`

Reviewer: We could explore extending the operator to support `myCnfSecretKeyRef`  in further iterations, to achieve this in a more native way.

Author: Ah, tunnel visioned here. I was literally working with Secret Mounts for ldap, yet forgot I can use them for this too. Thanks, that is perfect

        name: mariadb-vault
        key: my.cnf

      resources:
        requests:
          cpu: 100m
          memory: 128Mi
        limits:
          memory: 1Gi

      metrics:
        enabled: true
    ```
    `kubectl apply -f mariadb-vault.yaml`

6. **Verify Encryption Works.**
    ```sh
    kubectl run mariadb-connect --rm -it --image=mariadb:11.4 -- bash -c "mariadb -u root -p'MariaDB11!' --ssl=false -h mariadb"
    ```

    You should see something along the lines of:

    ```
    If you don't see a command prompt, try pressing enter.
    Welcome to the MariaDB monitor.  Commands end with ; or \g.
    Your MariaDB connection id is 95
    Server version: 11.4.7-4-MariaDB-enterprise MariaDB Enterprise Server

    Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

    Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

    MariaDB [(none)]>
    ```

    At this point, you can check the encryption status:
    ```sql
    SELECT * from information_schema.INNODB_TABLESPACES_ENCRYPTION;
    ```

    If you create a new database and then table, the above query should return additional information about them.

> ## Day-2 Ops

Reviewer: ```suggestion
## Day-2 Operations
```

> ### Rotating Secrets.

Reviewer: This is way better than what the file encryption plugin:
https://mariadb.com/resources/blog/mariadb-encryption-tde-using-mariadbs-file-key-management-encryption-plugin/

Do you see any benefit of supporting the file encryption plugin at this point?

Author: I feel like we are lacking in options. The file key management plugin is not production ready, while vault is great, but it's the **only** option right now.

At the same time, for some customers that file key management plugin could be a good middle ground.

Reviewer: We should raise this with Egor when delivering the docs.

1. **Put A New Secret In Vault.**
    After logging in to vault, you can run again:
    ```sh
    vault kv put /mariadb/1 data="$(openssl rand -hex 32)"
    vault kv put /mariadb/2 data="$(openssl rand -hex 32)"
    ```
    This will start re-encrypting data.

2. **Monitor Re-Encryption.**
    ```sh
    kubectl run mariadb-connect --rm -it --image=mariadb:11.4 -- bash -c "mariadb -u root -p'MariaDB11!' --ssl=false -h mariadb"
    ```
    If you check the encrpytion status again:
    ```sql

>     SELECT * from information_schema.INNODB_TABLESPACES_ENCRYPTION;

Reviewer: Could we share the output of this query?

Additionally, as described in [this blog](https://mariadb.com/resources/blog/how-to-implement-encryption-at-rest-using-hashicorp-vault-and-mariadb/), we can show that the `/var/lib/mysql/testdb/*.ibd` is actually encrypted.

Reviewer: It is clear the way it is

    ```

>     You should see `CURRENT_KEY_VERSION` column start getting updated to point to the new key version.

Reviewer: Show an example output of this `CURRENT_KEY_VERSION`

    ```sh
    cat <<'EOF' | vault policy write -non-interactive mariadb -
    # Allow access to MariaDB secrets
    path "mariadb/*" {
      capabilities = ["create", "read", "update", "delete", "list"]
    }

    # Allow reading the mount configuration
    path "sys/mounts/mariadb/tune" {
      capabilities = ["read"]
    }
    EOF
    ```
    After which, we can create a new token with the given policy.
    ```sh
    vault token create -policy mariadb
    ```
    You will see output similar to:
    ```
    Key                  Value
    ---                  -----
    token                EXAMPLE_TOKEN
    token_accessor       utFtmh98YAAJyYdxEVN3SFQA
    token_duration       768h
    token_renewable      true
    token_policies       ["default" "mariadb"]
    identity_policies    []
    policies             ["default" "mariadb"]
    ```
    Your new token is: `EXAMPLE_TOKEN`.

4. **Create A `Secret` For the vault token.**
    Now that you've either created a new token, or are using an existing one, we need to create a secret with it.
    ```sh
    export TOKEN="EXAMPLE_TOKEN"
    kubeclt create secret generic mariadb-vault-token --from-literal=token="$TOKEN"
    ```

> 5. **Create a Secret for the ca.**

Reviewer: ```suggestion
5. **Create a Secret for the Certificate Authority (CA) used to issue the Vault certificate.**
```

>     For which certificate is needed, consult [the docs](../../../server/security/securing-mariadb/encryption/data-at-rest-encryption/key-management-and-encryption-plugins/hashicorp-key-management-plugin.md#hashicorp-key-management-vault-ca)

Reviewer: ```suggestion
    For further information, consult [the docs](../../../server/security/securing-mariadb/encryption/data-at-rest-encryption/key-management-and-encryption-plugins/hashicorp-key-management-plugin.md#hashicorp-key-management-vault-ca)
```

    If you have the certificate locally in a file called `ca.crt` you can run:

    ```sh
    kubectl create secret generic vault-tls --from-file=./ca.crt
    ```

5. **Create A MariaDB Custom Resource.**
    The final step is creating a new MariaDB instance.

    **mariadb-vault.yaml**
    ```yaml
    ---
    apiVersion: v1
    kind: Secret
    metadata:
      name: mariadb # Used to hold the mariadb and root user passwords
      labels:

>         enterprise.mariadb.com/watch: ""

Author: Hmm, but in this case it should be ok, no? It's also part of the examples.

>         enterprise.mariadb.com/watch: ""

Reviewer: True, for one moment I got confused with the Vault token rotation

      rootPasswordSecretKeyRef:
        name: mariadb
        key: password

      username: mariadb
      passwordSecretKeyRef:
        name: mariadb-password
        key: password
        generate: true
      database: mariadb

      port: 3306

      storage:
        size: 1Gi
        # storageClassName: csi-hostpath-sc

      myCnf: |
        [mariadb]
        bind-address=*
        default_storage_engine=InnoDB
        binlog_format=row
        innodb_autoinc_lock_mode=2
        innodb_buffer_pool_size=800M
        max_allowed_packet=256M

        plugin_load_add = hashicorp_key_management
        hashicorp-key-management-vault-url=https://vault-0.vault-internal.default.svc.cluster.local:8200/v1/mariadb
        hashicorp-key-management-caching-enabled=ON
        hashicorp-key-management-vault-ca=/etc/vault/certs/ca.crt

        innodb_encrypt_tables = FORCE
        innodb_encrypt_log = ON
        innodb_encrypt_temporary_tables = ON
        encrypt_tmp_disk_tables = ON
        encrypt_tmp_files = ON
        encrypt_binlog = ON
        aria_encrypt_tables = ON

        innodb_encryption_threads = 4
        innodb_encryption_rotation_iops = 2000

      env:

>         - name: VAULT_TOKEN # This is where our token is defined!

Author: I have no idea, I mentioned it in one of the commits I believe but it's basically reverse engineering (in this case from a log that the plugin adds).

    | innodb_undo001  |                 1 |               1 |                   1 |              1 |                    0 |
    | innodb_undo002  |                 1 |               1 |                   1 |              1 |                    0 |
    | innodb_undo003  |                 1 |               1 |                   1 |              1 |                    0 |
    | mysql/innodb_ta |                 1 |               1 |                   1 |              1 |                    0 |
    | mysql/innodb_in |                 1 |               1 |                   1 |              1 |                    0 |
    | mysql/gtid_slav |                 1 |               1 |                   1 |              1 |                    0 |
    | mysql/transacti |                 1 |               1 |                   1 |              1 |                    0 |
    | my_db/people    |                 1 |               1 |                   1 |              1 |                    0 |
    +-----------------+-------------------+-----------------+---------------------+----------------+----------------------+
    ```

>     Note: The above query is truncated, so it will be readable in the docs. In reality, you will see a few more columns.

Reviewer: ```suggestion
    Note: The above query is truncated. In reality, you will see a few more columns.
```

    | innodb_undo003  |                 1 |               1 |                   2 |              1 |                    0 |
    | mysql/innodb_ta |                 1 |               1 |                   2 |              1 |                    0 |
    | mysql/innodb_in |                 1 |               1 |                   2 |              1 |                    0 |
    | mysql/gtid_slav |                 1 |               1 |                   2 |              1 |                    0 |
    | mysql/transacti |                 1 |               1 |                   2 |              1 |                    0 |
    | my_db/people    |                 1 |               1 |                   2 |              1 |                    0 |
    +-----------------+-------------------+-----------------+---------------------+----------------+----------------------+
    ```

### Rotating Token
Make sure when rotating the token, to do so in advance of the token expiring.

1. **Acquire a new token and update the secret.**

    ```sh
    export TOKEN="EXAMPLE_TOKEN"
    kubeclt create secret generic mariadb-vault-token --from-literal=token="$TOKEN"
    ```

> 2. **Restart The MariaDB StatefullSet.**

Reviewer: ```suggestion
2. **Restart The MariaDB StatefulSet.**
```

>     MariaDB will continue using the old token until the statefulset is restart. If the name of the database CRD is `mariadb` (as per the example above), then:

Reviewer: ```suggestion
    MariaDB will continue using the old token until the StatefulSet is restarted. If the name of the `MariaDB` CR  is `mariadb` (as per the example above), then:
```

    ```sh
    kubectl rollout restart statefulset mariadb
    ```

## Known Issues/Limitations

### **Vault Not Being Accessible Will Result In MariaDB Not Working**

> As MariaDB uses vault to fetch it's decryption key, in the case where vault is not accessible due to it being down, network issues, secret
> not existing, will result in MariaDB not being able to fetch the decryption key and hence stop working. While the Hashicorp plugin has a
> configurable cache, that should be set and will result in MariaDB still working for a few seconds to minutes, depending on configuration,
> the cache is not reliable as it's ephemeral and short lived.

Reviewer: ```suggestion
As MariaDB uses Vault to fetch it's decryption key, in case that Vault becomes unavailable, it will result in MariaDB not being able to fetch the decryption key and hence stop working. While the Hashicorp plugin has a configurable cache, that should be set and will result in MariaDB still working for a few seconds to minutes, depending on configuration, the cache is not reliable as it's ephemeral and short lived.
```
